ボウリングのスコア計算プログラム（オブジェクト指向版）を作成

uchihiro04 opened the pull request:
ボウリングのスコア計算プログラム（オブジェクト指向版）を作成しました。
Shotクラス、Frameクラス、Gameクラスの3つに分割しています。

uchihiro04 committed:
Shotクラスを作成

1投に関するクラスであるShotクラスを作成しました。

uchihiro04 committed:
Frameクラスを作成

フレームごとにインスタンスを作成できるように、Frameクラスを作成しました。

uchihiro04 committed:
得点を整数化するメソッドの名前を修正

得点を整数かするメソッドの名前をオブジェクト指向に沿うような形で修正しました。

uchihiro04 committed:
Gameクラスを作成

ボウリングのゲーム全体に関するクラスである、Gameクラスを作成しました。

uchihiro04 committed:
Gameクラスのインスタンスを作成するメソッドを追加

Frameクラスがフレーム分配列として入ったものを初期値とする、Gameクラスのインスタンスを作成するメソッドを追加しました。

uchihiro04 committed:
ボウリングのスコアの出力を実装

ボウリングのスコアの出力を実装しました。

uchihiro04 committed:
Shotクラスで定義していたメソッドを修正

Shotクラスのメソッド名を変更したため、Frameクラス内で使用している箇所について修正を行いました。

uchihiro04 committed:
メソッドの追加・メソッド名の変更

ボウリングのスコアを計算するメソッドを追加しました。また、Gameクラスのインスタンスを作成するメソッドの名前を変更しました。

uchihiro04 committed:
テストを作成

ボウリングのスコア計算について数パターンテストを作成しました。

uchihiro04 committed:
テスト実行時に発生していたエラーを修正

テスト実行時にエラーが発生していた箇所を修正しました。

uchihiro04 committed:
メソッドを分割

1つのメソッドを、スコアのかたまりをフレームごとに分割するメソッドと、Gameクラスのインスタンスを作成するメソッドに分割しました。

uchihiro04 committed:
ストライク・スペアを判断するメソッドをそれぞれ作成

ストライク・スペアかどうか判断するメソッドをそれぞれ作成しました。

uchihiro04 committed:
calc_scoresメソッドの一部の条件分岐を削除

calc_scoresメソッドにて使われていた一部の条件分岐を削除しました。

uchihiro04 committed:
デバッグ用に記述していたrequireを削除

デバッグ用に記述していたrequireを削除しました。

uchihiro04 committed:
attr_readerの削除

不要なattr_readerを削除し、コードを修正しました。

uchihiro04 committed:
不要なattr_readerを一部削除

不要なattr_readerを一部削除しました。

uchihiro04 committed:
Gameクラスのインスタンスの初期値@framesに、Frameクラスのインスタンスを代入するメソッドを作成

Gameクラスのインスタンスの初期値である@framesに、Frameクラスのインスタンスの配列を代入するメソッドを作成しました。また、メソッドの作成に伴いcalc_scoresメソッドを修正しました。

uchihiro04 committed:
テストを追加

テストに「ストライクを取った時のパターン」「スペアを取ったときのパターン」の2種類を追加しました。

uchihiro04 committed:
split_scoresメソッド内のコードを一部削除

split_scoresメソッド内のコードを一部削除しました。

uchihiro04 committed:
一部メソッドをprivateメソッドに変更

Gameクラス内でのみ使われていた一部メソッドをprivateメソッドに変更しました、

uchihiro04 committed:
rubocopの指摘点を修正

rubocopからの指摘を受け、コードの修正およびメソッドの分割を行いました。

uchihiro04 committed:
Frameクラスのinitializeメソッドを変更

Frameクラスのinitializeメソッド実行時の初期値について、データ型をstringからintegerになるように変更しました。

uchihiro04 committed:
不要な空行を削除

rubocopに指摘された不要な空行を削除しました。

u1tnk commented on 07.bowling_object/lib/frame.rb:
```diff
@@ -0,0 +1,25 @@
+  attr_reader :first_shot, :second_shot
+
+  def initialize(first_mark, second_mark = nil, third_mark = nil)
+    @first_shot = Shot.new(first_mark).score
```
まず、Shotインスタンスはscoreを出す機能があるので情報量的にFrameはShotクラスのインスタンスを保持するべきです。
また、 変数名として`xxx_shot`とあるとShotクラスのインスタンスかな？というので中身が数値になってしまっているのはよくないというのもあります。

uchihiro04 replied:
Frameクラスのインスタンスの`@first_shot`、`@second_shot``@third_shot`について、Shotクラスのインスタンスを保持するように修正いたしました。

該当commitは[0512a26](https://github.com/uchihiro04/ruby-practices/pull/10/commits/0512a26003812c03ec99d67cc8aa08651a8a650a)です。

u1tnk commented on 07.bowling_object/lib/game.rb:
```diff
@@ -0,0 +1,58 @@
+
+class Game
+  def initialize(scores)
+    @frames = []
```
framesのデータはメソッドを呼んだものを使うのでここで配列だけ初期化しているのは違和感があります。
framesメソッドを initialize_frames等の名前にしてここで呼び出して`@frames`を初期化するのが良いかと思います。

uchihiro04 replied:
Frameクラスのインスタンスにつきましては、Gameクラスのインスタンスメソッドを使って作成する方法から、Frameクラスのクラスメソッドを使って作成する方法に切り替えました。
ご提案いただいた方法とは異なる形になっております。ご確認いただけますと幸いです🙇‍♂️

該当するcommit

u1tnk reviewed (changes requested):
クラス構成も良いし、概ねシンプルな実装で良いのですが、Gameクラスに偏り過ぎだと思います。
コメントしたところ以外はFrameクラスに処理をいくらか渡す事ができれば良いかと思います 💪

uchihiro04 committed:
Frameクラスのインスタンスが保持するデータを数値からShotクラスのインスタンスに変更

Frameクラスのインスタンスが保持するデータを数値からShotクラスのインスタンスに変更しました。また、影響のあるFrameクラス、Gameクラスのメソッドを修正しました。

uchihiro04 committed:
Gameクラスのinitializeメソッドを変更

Gameクラスのinitializeメソッドについて、インスタンス変数framesが保持するデータを空の配列からFrameクラスのインスタンスに変更しました。

uchihiro04 committed:
Frameクラスのインスタンスの配列を作成するメソッドを、Frameクラスに移動

Frameクラスのインスタンスの配列を作成するメソッドを、GameクラスのインスタンスメソッドからFrameクラスのクラスメソッドに移動しました。

uchihiro04 committed:
不要なメソッドを削除

使われなくなっていたメソッドを削除しました。

uchihiro04 committed:
ストライク・スペア時のボーナスに関わるメソッドをGameクラスからFrameクラスに移動

ストライク・スペア時のボーナスに関わるメソッドをGameクラスからFrameクラスに移動しました。

uchihiro04 committed:
privateメソッドを設定

クラス外で使用しないメソッドについて、privateメソッドを設定しました。

uchihiro04 committed:
メソッドのレシーバおよび引数を修正

メソッドのレシーバについてインスタンス変数で問題ない箇所はインスタンス変数に修正しました。また、引数が一部必要なくなったメソッドを修正しました。

uchihiro04 committed:
メソッドのレシーバを修正

メソッドのレシーバを一部、インスタンス変数に修正しました。

uchihiro04 committed:
メソッドの不要になった引数を削除

next_frameメソッド、next_next_frameメソッドについて、不要になった引数を削除しました。

uchihiro04 committed:
Frameクラスのメソッドの引数名を変更

Frameクラスのインスタンスメソッドの引数名を適した形に変更しました。

u1tnk commented on 07.bowling_object/lib/game.rb:
```diff
@@ -0,0 +1,21 @@
+
+  def calc_scores
+    @frames.each.sum do |frame|
+      if frame.strike? && frame.index < 9
```
ここまで来たら `Frame#score`を作って、bonusも含めたそのFrameのスコアを返してあげて、ここでは足すだけにしましょう！

uchihiro04 replied:
Frameクラスに、ボーナス点も含めたフレームのスコアを返すscoreメソッドを作成しました。

該当するcommitは[e688996](https://github.com/uchihiro04/ruby-practices/pull/10/commits/e68899660521fe5454e18dda9b6caf2ff83f47c4)となります。

u1tnk reviewed (changes requested):
あと一点対応頂ければ大丈夫です！もうちょと 💪

uchihiro04 committed:
フレームのスコア計算をGameクラスからFrameクラスのメソッドに移行

Gameクラスのcalc_scoresメソッド内で行っていたフレームのスコア計算を、Frameクラスのscoreメソッドに移行しました。

uchihiro04 committed:
不要なアクセサメソッドを削除

使われていないアクセサメソッドを削除しました。

u1tnk commented on 07.bowling_object/lib/frame.rb:
```diff
@@ -23,6 +23,16 @@ def self.generate_frames(scores)
     frames.map.with_index { |frame, index| Frame.new(index, frame[0], frame[1], frame[2]) }
   end
 
+  def score(frames, frame)
```
frame.score(@frames, frame)惜しい！ そのまま移動しちゃったんだと思いますが、`frame.score(@frames, frame)` の frame二つは同じものだと思うので引数で渡す必要無いですよ。
このメソッド内では引数では渡さず、替わりに省略すれば良いです。

ここで `strike?`を呼べば 内部的には `self.strike?`と同義で selfがframeになります。
ちなみに今回無いと思いますが 例えば 、`first_shot = shot`みたいに代入が必要なときは変数と区別できないので省略できなず、 `self.first_shot = shot`みたいにする必要があるのでご注意ください。

uchihiro04 replied:
`frame.score(@Frames, frame)`から引数`frame`を削除し、あわせて`Frame#score`メソッド内の`frame`を削除しました。

該当するcommitは[6c0489f](https://github.com/uchihiro04/ruby-practices/pull/10/commits/6c0489f8c6822e79bf8c13ae09feb9b49321d46c)となります。

u1tnk reviewed (changes requested):
もうほんのちょっとです！

最後に必要無くなったアクセサを消してるのはすごく良いですね 👍 
忘れがちだけど大事です。

uchihiro04 committed:
Frame#scoreメソッドの引数を一部削除

不要になっていたFrame#scoreメソッドの一部の引数を削除しました。

u1tnk reviewed (approved):
👍👍👍
